core/src/main/java/org/elasticsearch/action/bulk/BulkItemRequest.java

```diff
@@ -94,15 +94,36 @@ public void readFrom(StreamInput in) throws IOException {
         request = DocWriteRequest.readDocumentRequest(in);
         if (in.readBoolean()) {
             primaryResponse = BulkItemResponse.readBulkItem(in);
+            if (in.getVersion().onOrAfter(Version.V_5_6_0_UNRELEASED)) {
+                // 6.0 no longer mutates the requests with these, do it here as a bwc layer
+                request.version(primaryResponse.getVersion());
+                request.versionType(request.versionType().versionTypeForReplicationAndRecovery());
+            }
+        }
+        if (in.getVersion().before(Version.V_5_6_0_UNRELEASED)) {
+            ignoreOnReplica = in.readBoolean();
+            if (ignoreOnReplica == false && primaryResponse != null) {
+                assert primaryResponse.isFailed() == false : "expected no failure on the primary response";
+            }
+        } else {
+            ignoreOnReplica = primaryResponse != null &&
+                (primaryResponse.isFailed() || primaryResponse.getResponse().getResult() == DocWriteResponse.Result.NOOP);
         }
-        ignoreOnReplica = in.readBoolean();
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(id);
         DocWriteRequest.writeDocumentRequest(out, request);
         out.writeOptionalStreamable(primaryResponse);
-        out.writeBoolean(ignoreOnReplica);
+        assert ignoreOnReplica ==
```

Comment:
It feels weird to place this assert here, rather than where `ignoreOnReplica` is actually set, it also would make debugging more difficult, because instead of a stacktrace letting you know where it was incorrectly set, it only manifests once serialized.

How would you feel about combining `setPrimaryResponse` and `setIgnoreOnReplica` so they are entangled and an exception is thrown if an invalid state is configured when the primary response is set? Perhaps something like:

```java
void setPrimaryResponse(BulkItemResponse primaryResponse, boolean ignoreOnReplica) {
    assert ignoreOnReplica ==
        (primaryResponse != null &&
            (primaryResponse.isFailed() || primaryResponse.getResponse().getResult() == DocWriteResponse.Result.NOOP)
        ) :
        "unexpected ignoreOnReplica value. primaryResponse [" + primaryResponse + "], primaryResponse ["
            + (primaryResponse == null ? "null" : XContentHelper.toString(primaryResponse)) + "]";
    this.primaryResponse = primaryResponse;
    this.ignoreOnReplica = ignoreOnReplica;
}
```

It could also be configured as a regular exception there also, instead of an assert (`IllegalArgumentException`). It would also have the benefit of not leaving the object in an illegal state when an exception was thrown (nice to be clean)